spec/models/member_spec.rb

```diff
@@ -160,4 +163,142 @@
+      end
+
+      it "define the end_subscription date one year after the day of subscription when the year is a multiple of 400" do
+        allow(Date).to receive(:today) { Date.new(2399, 10, 24) }
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2400, 10, 23))
+      end
+
+      it "define the end_subscription date one year after the day of subscription when the day is 29 february" do
+        allow(Date).to receive(:today) { Date.new(2060, 2, 29) }
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2061, 2, 28))
+      end
+    end
+
+    context "when is subscription is still active" do
+      let(:member) { create :member }
+
+      it "define the end_subscription date one year after the day of subscription" do
+        arr_date = []
+        arr_date << [Date.new(2001, 2, 5), Date.new(2001, 4, 7), Date.new(2002, 4, 7)]
+        arr_date << [Date.new(2121, 7, 21), Date.new(2121, 10, 5), Date.new(2122, 10, 5)]
+        arr_date << [Date.new(3014, 5, 5), Date.new(3015, 2, 2), Date.new(3016, 2, 2)]
+
+        arr_date.each do |tab|
+          allow(Date).to receive(:today) { tab[0] }
+          member.end_subscription = tab[1]
+          member.renew_subscription_date
+          expect(member.end_subscription).to eq(tab[2])
+        end
+      end
+
+      it "define the end_subscription date one year and one day after the actual end_subscription date when the is a multiple of 100" do
+        allow(Date).to receive(:today) { Date.new(2099, 11, 11) }
+        member.end_subscription = Date.new(2100, 1, 11)
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2101, 1, 11))
+      end
+
+      it "define the end_subscription date one year and one day after the actual end_subscription date when the year is a leapyear" do
+        allow(Date).to receive(:today) { Date.new(2004, 2, 5) }
+        member.end_subscription = Date.new(2004, 2, 12)
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2005, 2, 12))
+      end
+
+      it "define the end_subscription date one year and one day after the actual end_subscription date when the year is a multiple of 400" do
+        allow(Date).to receive(:today) { Date.new(2399, 10, 24) }
+        member.end_subscription = Date.new(2400, 2, 9)
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2401, 2, 9))
+      end
+
+      it "define the end_subscription date one year and one day after the actual end_subscription date when the end_subscription date is a 29 fubruary" do
+        allow(Date).to receive(:today) { Date.new(2024, 1, 24) }
+        member.end_subscription = Date.new(2024, 2, 29)
+        member.renew_subscription_date
+        expect(member.end_subscription).to eq(Date.new(2025, 2, 28))
+      end
+    end
+  end
```

Comment:
For each of the `context` that you have set, the tests structure are the same each time, and the tests types are also the same. Only the initial conditions differ. You should try to refactor this file to make it DRYer.
This could be a situation where you could use shared_examples, but since this kind of tests appears only in this file, I think it'll be better if you just create a method to call these tests with the initial conditions as arguments.

First, refactor small things. Then when the 'story' becomes simpler to read, you could find a way to extract each context.

Reply:
done